<a href="https://colab.research.google.com/github/beatrizjafelice/Association-Rules-for-Artists/blob/master/Association_Rules_for_Artists_PT.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Atividade
1. Montar uma base colaborativa em que cada aluno deve informar, em uma mesma linha, separados por vírgula, os nomes de suas bandas favoritas (min: 2 e max: 5). Por questões de padronizaçao, os nomes devem estar em letra maiúscula e com _ no lugar dos espaços.

2. Extrair regras de associação da base montada.

3. Selecionar as regras que se aplicam a você cujos consequentes (bandas) você não conhece. Ouça uma música de cada e anote o resultado :)

Extraindo os dados:

In [ ]:
import pandas as pd

df = pd.read_csv('https://drive.google.com/u/3/uc?id=1aDEAX-LqUtdLl10n5SrjqF-dMAmDBmTE&export=download', names=['bandas'])
df.head()

bandas
0  MAMONAS_ASSASSINAS, QUEEN, RACIONAIS_MCS, EXAL...
1  COLDPLAY, THE_KILLERS, U2, ARCTIC_MONKEYS, ONE...
2  DREAM_THEATER, AVENGED_SEVENFOLD, OFICINA_G3, ...
3             METALLICA, TERNO_REI, PARAMORE, ZIMBRA
4  IMAGINE_DRAGONS, GUNS_AND_ROSES, COLDPLAY, LUM...

Em análises de regras de associação, as linhas do conjunto de dados são chamadas de "transações". Cada transação é composta por um ou mais itens. Nesse caso, as transações representam os artistas favoritos dos alunos e cada item corresponde a um artista.

Separando os itens em listas:

In [ ]:
data = list(df['bandas'].apply(lambda x: x.split(',')))
data

[['MAMONAS_ASSASSINAS', ' QUEEN', ' RACIONAIS_MCS', ' EXALTASSAMBA'],
 ['COLDPLAY',
  ' THE_KILLERS',
  ' U2',
  ' ARCTIC_MONKEYS',
  ' ONE_REPUBLIC',
  ' IMAGINE_DRAGONS'],
 ['DREAM_THEATER', ' AVENGED_SEVENFOLD', ' OFICINA_G3', ' ROSA_DE_SARON'],
 ['METALLICA', ' TERNO_REI', ' PARAMORE', ' ZIMBRA'],
 ['IMAGINE_DRAGONS', ' GUNS_AND_ROSES', ' COLDPLAY', ' LUMINNIERS'],
 ['RACIONAIS_MCS', ' DJONGA', ' JEAN_TASSY', ' PURO_SUCO'],
 ['JOURNEY',
  ' SUPERTRAMP',
  ' ABBA',
  ' EARTH_WIND_AND_FIRE',
  ' TEARS_FOR_FEARS'],
 ['QUEEN', ' BLONDIE', ' MUSE', ' PINK_FLOYD', ' IMAGINE_DRAGONS'],
 ['CHARLIE_BROWN_JR',
  ' MENOS_E_MAIS',
  ' EXALTASSAMBA',
  ' RACIONAIS_MCS',
  ' NATIRUTS',
  ' ONZE:20',
  ' FALAMANSA']]

Removendo espaços extras para evitar duplicidades:

In [ ]:
for lista in data:
    for i, nome in enumerate(lista):
        lista[i] = nome.strip()
data

[['MAMONAS_ASSASSINAS', 'QUEEN', 'RACIONAIS_MCS', 'EXALTASSAMBA'],
 ['COLDPLAY',
  'THE_KILLERS',
  'U2',
  'ARCTIC_MONKEYS',
  'ONE_REPUBLIC',
  'IMAGINE_DRAGONS'],
 ['DREAM_THEATER', 'AVENGED_SEVENFOLD', 'OFICINA_G3', 'ROSA_DE_SARON'],
 ['METALLICA', 'TERNO_REI', 'PARAMORE', 'ZIMBRA'],
 ['IMAGINE_DRAGONS', 'GUNS_AND_ROSES', 'COLDPLAY', 'LUMINNIERS'],
 ['RACIONAIS_MCS', 'DJONGA', 'JEAN_TASSY', 'PURO_SUCO'],
 ['JOURNEY', 'SUPERTRAMP', 'ABBA', 'EARTH_WIND_AND_FIRE', 'TEARS_FOR_FEARS'],
 ['QUEEN', 'BLONDIE', 'MUSE', 'PINK_FLOYD', 'IMAGINE_DRAGONS'],
 ['CHARLIE_BROWN_JR',
  'MENOS_E_MAIS',
  'EXALTASSAMBA',
  'RACIONAIS_MCS',
  'NATIRUTS',
  'ONZE:20',
  'FALAMANSA']]

Agora, codifica-se o conjunto de dados como uma tabela atributo-valor. Nessa tabela, cada linha representa uma transação e cada coluna representa um item. Os valores True ou False indicam se o item da coluna está presente naquele registro ou não.

In [ ]:
from mlxtend.preprocessing import TransactionEncoder

t_enc = TransactionEncoder()
t_enc_data = t_enc.fit(data).transform(data)
t_enc_data

array([[False, False, False, False, False, False, False, False, False,
         True, False, False, False, False, False, False,  True, False,
        False, False, False, False, False, False, False, False, False,
         True,  True, False, False, False, False, False, False, False],
       [False,  True, False, False, False,  True, False, False, False,
        False, False, False,  True, False, False, False, False, False,
        False, False, False, False,  True, False, False, False, False,
        False, False, False, False, False, False,  True,  True, False],
       [False, False,  True, False, False, False, False,  True, False,
        False, False, False, False, False, False, False, False, False,
        False, False, False,  True, False, False, False, False, False,
        False, False,  True, False, False, False, False, False, False],
       [False, False, False, False, False, False, False, False, False,
        False, False, False, False, False, False, False, False, False,
   

Transformando em um dataframe para melhor visualização:

In [ ]:
df_encoded = pd.DataFrame(t_enc_data, columns=t_enc.columns_)
df_encoded

ABBA  ARCTIC_MONKEYS  AVENGED_SEVENFOLD  BLONDIE  CHARLIE_BROWN_JR  \
0  False           False              False    False             False   
1  False            True              False    False             False   
2  False           False               True    False             False   
3  False           False              False    False             False   
4  False           False              False    False             False   
5  False           False              False    False             False   
6   True           False              False    False             False   
7  False           False              False     True             False   
8  False           False              False    False              True   

   COLDPLAY  DJONGA  DREAM_THEATER  EARTH_WIND_AND_FIRE  EXALTASSAMBA  ...  \
0     False   False          False                False          True  ...   
1      True   False          False                False         False  ...   
2     False   False           True                False         False  ...   
3     False   False          False                False         False  ...   
4      True   False          False                False         False  ...   
5     False    True          False                False         False  ...   
6     False   False          False                 True         False  ...   
7     False   False          False                False         False  ...   
8     False   False          False                False          True  ...   

   PURO_SUCO  QUEEN  RACIONAIS_MCS  ROSA_DE_SARON  SUPERTRAMP  \
0      False   True           True          False       False   
1      False  False          False          False       False   
2      False  False          False           True       False   
3      False  False          False          False       False   
4      False  False          False          False       False   
5       True  False           True          False       False   
6      False  False          False          False        True   
7      False   True          False          False       False   
8      False  False           True          False       False   

   TEARS_FOR_FEARS  TERNO_REI  THE_KILLERS     U2  ZIMBRA  
0            False      False        False  False   False  
1            False      False         True   True   False  
2            False      False        False  False   False  
3            False       True        False  False    True  
4            False      False        False  False   False  
5            False      False        False  False   False  
6             True      False        False  False   False  
7            False      False        False  False   False  
8            False      False        False  False   False  

[9 rows x 36 columns]

Em seguida, extrai-se os *itemsets*, que são os conjuntos de itens que aparecem com mais frequência na base. Os *itemsets* podem ser formados por um ou mais itens. O parâmetro *min_support* determina a frequência mínima do itemset na base para ser considerado frequente.

In [ ]:
from mlxtend.frequent_patterns import apriori, association_rules

# OBS: em bases de dados grandes, aumentar o min_support
# (mínima proporção de transações em que o itemset aparece)
freq_itemsets = apriori(df_encoded, min_support=0.2, use_colnames=True)
freq_itemsets

support                       itemsets
0  0.222222                     (COLDPLAY)
1  0.222222                 (EXALTASSAMBA)
2  0.333333              (IMAGINE_DRAGONS)
3  0.222222                        (QUEEN)
4  0.333333                (RACIONAIS_MCS)
5  0.222222    (IMAGINE_DRAGONS, COLDPLAY)
6  0.222222  (EXALTASSAMBA, RACIONAIS_MCS)

Extraindo as regras de associação:

In [ ]:
regras = association_rules(freq_itemsets, metric='confidence', min_threshold=0.6)
regras

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  def _forward_input(self, allow_stdin=False):


antecedents        consequents  antecedent support  \
0  (IMAGINE_DRAGONS)         (COLDPLAY)            0.333333   
1         (COLDPLAY)  (IMAGINE_DRAGONS)            0.222222   
2     (EXALTASSAMBA)    (RACIONAIS_MCS)            0.222222   
3    (RACIONAIS_MCS)     (EXALTASSAMBA)            0.333333   

   consequent support   support  confidence  lift  leverage  conviction  \
0            0.222222  0.222222    0.666667   3.0  0.148148    2.333333   
1            0.333333  0.222222    1.000000   3.0  0.148148         inf   
2            0.333333  0.222222    1.000000   3.0  0.148148         inf   
3            0.222222  0.222222    0.666667   3.0  0.148148    2.333333   

   zhangs_metric  
0       1.000000  
1       0.857143  
2       0.857143  
3       1.000000

A tabela de regras de associação apresenta relações entre artistas e bandas, mostrando com que frequência eles são ouvidos juntos. Aqui está uma interpretação detalhada de cada coluna na tabela:

1. **Antecedents (Antecedente)**: O artista ou banda que aparece em primeiro lugar na regra.
2. **Consequents (Consequente)**: O artista ou banda que aparece em segundo lugar na regra.
3. **Antecedent support**: A proporção de transações que incluem o antecedente.
4. **Consequent support**: A proporção de transações que incluem o consequente.
5. **Support (Suporte)**: A proporção de transações que incluem tanto o antecedente quanto o consequente.
6. **Confidence (Confiança)**: A probabilidade de o consequente ser ouvido quando o antecedente é ouvido.
7. **Lift**: A razão entre a confiança da regra e o suporte do consequente. Compara o suporte observado e o suporte esperado se o antecedente e o consequente fossem independentes. Um *lift* maior que 1 indica uma associação positiva.
8. **Leverage**: A diferença entre o suporte observado e o suporte esperado se o antecedente e o consequente fossem independentes. Valores próximos de 1 indicam associação positiva, próximos de 0 indicam nenhuma associação e próximos de -1, associação negativa.
9. **Conviction**: Relação entre a frequência esperada do antecedente sem o consequente e a frequência observada do antecedente sem o consequente. Indica quanto mais frequentemente o antecedente implica o consequente em comparação com os casos onde o antecedente não implica o consequente. Valores altos indicam associação positiva.
10. **Métrica de Zhang**: Analisa como a presença e a não presença do antecedente impactam a presença do consequente. Pode assumir valores entre -1 e 1, assim como *leverage*.

### Interpretação das Regras

1. **Regra 1: (COLDPLAY) -> (IMAGINE_DRAGONS)**
   - **Suporte do antecedente**: 0.222222 (22.22% das transações incluem COLDPLAY)
   - **Suporte do consequente**: 0.333333 (33.33% das transações incluem IMAGINE_DRAGONS)
   - **Suporte**: 0.222222 (22.22% das transações incluem tanto COLDPLAY quanto IMAGINE_DRAGONS)
   - **Confidence**: 1.000000 (100% das transações que incluem COLDPLAY também incluem IMAGINE_DRAGONS)
   - **Lift**: 3.0 (Ouvintes de COLDPLAY têm 3 vezes mais chances de ouvir IMAGINE_DRAGONS do que seria esperado por acaso)
   - **Leverage**: 0.148148 (Valor positivo indica uma associação positiva)
   - **Conviction**: inf (Infinita, sugerindo uma implicação muito forte, já que COLDPLAY sempre leva a IMAGINE_DRAGONS)
   - **Métrica de Zhang**: 0.857143 (Valor alto indicando uma associação forte)

2. **Regra 2: (IMAGINE_DRAGONS) -> (COLDPLAY)**
   - **Suporte do antecedente**: 0.333333 (33.33% das transações incluem IMAGINE_DRAGONS)
   - **Suporte do consequente**: 0.222222 (22.22% das transações incluem COLDPLAY)
   - **Suporte**: 0.222222 (22.22% das transações incluem tanto IMAGINE_DRAGONS quanto COLDPLAY)
   - **Confidence**: 0.666667 (66.67% das transações que incluem IMAGINE_DRAGONS também incluem COLDPLAY)
   - **Lift**: 3.0 (Ouvintes de IMAGINE_DRAGONS têm 3 vezes mais chances de ouvir COLDPLAY do que seria esperado por acaso)
   - **Leverage**: 0.148148 (Valor positivo indica uma associação positiva)
   - **Conviction**: 2.333333 (Indica que a regra IMAGINE_DRAGONS -> COLDPLAY é bastante forte, mas não tão forte quanto a regra inversa)
   - **Métrica de Zhang**: 1.000000 (Valor alto indicando uma associação forte)

3. **Regra 3: (EXALTASSAMBA) -> (RACIONAIS_MCS)**
   - **Suporte do antecedente**: 0.222222 (22.22% das transações incluem EXALTASSAMBA)
   - **Suporte do consequente**: 0.333333 (33.33% das transações incluem RACIONAIS_MCS)
   - **Suporte**: 0.222222 (22.22% das transações incluem tanto EXALTASSAMBA quanto RACIONAIS_MCS)
   - **Confidence**: 1.000000 (100% das transações que incluem EXALTASSAMBA também incluem RACIONAIS_MCS)
   - **Lift**: 3.0 (Ouvintes de EXALTASSAMBA têm 3 vezes mais chances de ouvir RACIONAIS_MCS do que seria esperado por acaso)
   - **Leverage**: 0.148148 (Valor positivo indica uma associação positiva)
   - **Conviction**: inf (Infinita, sugerindo uma implicação muito forte, já que EXALTASSAMBA sempre leva a RACIONAIS_MCS)
   - **Métrica de Zhang**: 0.857143 (Valor alto indicando uma associação forte)

4. **Regra 4: (RACIONAIS_MCS) -> (EXALTASSAMBA)**
   - **Suporte do antecedente**: 0.333333 (33.33% das transações incluem RACIONAIS_MCS)
   - **Suporte do consequente**: 0.222222 (22.22% das transações incluem EXALTASSAMBA)
   - **Suporte**: 0.222222 (22.22% das transações incluem tanto RACIONAIS_MCS quanto EXALTASSAMBA)
   - **Confidence**: 0.666667 (66.67% das transações que incluem RACIONAIS_MCS também incluem EXALTASSAMBA)
   - **Lift**: 3.0 (Ouvintes de RACIONAIS_MCS têm 3 vezes mais chances de ouvir EXALTASSAMBA do que seria esperado por acaso)
   - **Leverage**: 0.148148 (Valor positivo indica uma associação positiva)
   - **Conviction**: 2.333333 (Indica que a regra RACIONAIS_MCS -> EXALTASSAMBA é bastante forte, mas não tão forte quanto a regra inversa)
   - **Métrica de Zhang**: 1.000000 (Valor alto indicando uma associação forte)

### Resumo

- Os pares (COLDPLAY, IMAGINE_DRAGONS) e (EXALTASSAMBA, RACIONAIS_MCS) mostram associações fortes, indicadas por altos valores para *confidence*, *lift* e *leverage*.
- As regras onde o antecedente prediz perfeitamente o consequente (*confidence* de 1.000000) mostram um valor infinito para *conviction*, o que significa que o consequente está sempre presente quando o antecedente está presente.
- Os altos valores de *lift* (todos 3.0) sugerem que esses artistas/bandas têm três vezes mais chances de serem ouvidos juntos do que por acaso.
- Os valores da métrica de Zhang próximos de 1 ou altos (0.857143 a 1.000000) confirmam ainda mais a força dessas associações.
<br>
<br>

**Observação:** o dataset é pequeno e as regras extraídas são observadas apenas nos dados fornecidos. Portanto, não refletem uma tendência geral.